### PR TITLE
Add initial support for an escape character #2354

### PR DESCRIPTION
**Pull request by Filip62**
## What type of PR is this? (check all applicable)

* [x] Feature
* [x] Documentation Update
* [ ] Refactor
* [ ] Bug Fix
* [ ] Optimization
* [ ] Other: *Replace this with a description of the type of this PR*

## Description

* Introduce an escape character '\\'.
* Warn users about unescaped characters (only backslashes for now).
* For now, treat a single unescaped backslash as a literal backslash character.
* Treat two consecutive backslashes as a single literal backslash character.

## Related Issues & Documents
Implements first step of #2354

## Documentation (check all applicable)

* [x] This PR requires changes to the Wiki documentation (describe the changes)
See #2354  
* [ ] This PR requires changes to the documentation inside the git repo (please add them to the PR).
* [ ] Does not require documentation changes

### CODE REVIEW

**Review by patrick96 — CHANGES_REQUESTED** [review]
Thanks :)

I think everything works, good job!

As far as I can tell you are using `positive_offset` and `negative_offset` to keep track of the difference in indices between `cfg_value` and the returned value (and `escaped_value`).

Now, I don't think it's necessary to generate the escaped string for when there are unescaped backslashes. In that case, it would be possible to use the same string object for searching and replacing without having to keep track of index differences.

There are also a few other things that should be in this PR:

* The manpage for the polybar config should be updated to include information about the escape character. The file is in `doc/man/polybar.5.rst` and the changes should go in the `Keys` section.
* New tests for the new function as well as some added testcases for the `ParseLineKeyTest` that include backslashes. The tests belong in `tests/unit_tests/components/config_parser.cpp`
* A changelog entry describing your changes, including breaking changes: https://github.com/polybar/polybar/blob/master/CONTRIBUTING.md#changelog

File: src/components/config_parser.cpp
Comment (Filip62) [reply src/components/config_parser.cpp]: I am not sure how exactly should it look,  but I will do at least something.

**Comment by patrick96** [comment]
Let me know once this is ready for review again.

Also, the CI docs fail isn't your fault, there seems to be an issue with an ubuntu mirror.

File: src/components/config_parser.cpp
Comment (patrick96) [inline src/components/config_parser.cpp:308]: You have some weird indentation here (2 spaces in the if branch and 1 in the else branch), please use 2 spaces.

Also setting the `backslash_pos` is the same in both branches, you can move it after the if-else block.

File: src/components/config_parser.cpp
Comment (patrick96) [inline src/components/config_parser.cpp:302]: This can still lead to an out-of-bounds access. Maybe this is the cause of the segfault.

The special case happens only if there is an unescaped backslash at the end of the line, the behavior should be the same as if there was an unescaped backslash anywhere else.

File: tests/unit_tests/components/config_parser.cpp
Comment (patrick96) [inline tests/unit_tests/components/config_parser.cpp:255]:
```suggestion
 * The first element of the pair is the expected value and the second
```

**Review by patrick96 — CHANGES_REQUESTED** [review]
Code looks good now. Just some minor adjustments.

Also, I'm curious why you are using `find_first_of` instead of `find`

**Comment by patrick96** [comment]
Also, it seems that CI is also having a segfault. I was also able to reproduce it sometimes. Looking at the coredump, it looks like it's something inside the logger, but the actual errors are kind of weird. I suspect that it's because you used string concatenation for the log message.

File: src/components/config_parser.cpp
Comment (Filip62) [reply src/components/config_parser.cpp:314]: Clangd tells me that it will be copied if I remove the move call.

**Comment by Filip62** [comment]
> Code looks good now. Just some minor adjustments.
> 
> Also, I'm curious why you are using `find_first_of` instead of `find`

No reason, changed it to `find`.

> Also, it seems that CI is also having a segfault. I was also able to reproduce it sometimes. Looking at the coredump, it looks like it's something inside the logger, but the actual errors are kind of weird. I suspect that it's because you used string concatenation for the log message.

Printf tokens didn't help.

File: src/components/config_parser.cpp
Comment (Filip62) [reply src/components/config_parser.cpp]: Could you please explain why?
I added a fix if I understand what you meant, but the segfaults are still there.

File: src/components/config_parser.cpp
Comment (patrick96) [reply src/components/config_parser.cpp:318]: In that case, leave it. I'm sure the compiler know more than me ;)

**Review by patrick96 — CHANGES_REQUESTED** [review]
I think I found the issue with the segfault. It was undefined behavior in test code I wrote.

I pushed a fix in https://github.com/polybar/polybar/pull/2364.

After that is merged we can see if there are any other CI issues with your code.
But other than that, I think we can merge this soon.

File: src/components/config_parser.cpp
Comment (patrick96) [reply src/components/config_parser.cpp]: You're change is correct. Also, this was not the cause of the segfault.

**Comment by patrick96** [comment]
Can you rebase against master so that your branch includes my fix?

File: src/components/config_parser.cpp
Comment (Filip62) [reply src/components/config_parser.cpp:305]: Yeah I was just testing.

**Comment by Filip62** [comment]
If we remove the `backslash_pos == value.size() - 1` check, why would it lead to out of bounds access?
Wouldn't it be just looking at the string termination character?

**Review by patrick96 — APPROVED** [review]

**Comment by patrick96** [comment]
Everything looks good now, but I can't merge yet because there is still a merge conflict (though it wasn't there a minute ago). You need to resolve it before I can merge this.

**Comment by Filip62** [comment]
> Everything looks good now, but I can't merge yet because there is still a merge conflict (though it wasn't there a minute ago). You need to resolve it before I can merge this.

I am not sure if I am missing something but `git merge upstream/master` tells me I am up to date.

**Comment by patrick96** [comment]
I think you need to to `git rebase upstream/master`.

**Comment by patrick96** [comment]
I think https://github.com/polybar/polybar/pull/2361/commits/6b21044529e2344861abcba3b57b658efaa921da messed things up a bit

**Comment by Filip62** [comment]
Sorry for breaking things, still get confused by git sometimes.

**Comment by patrick96** [comment]
There was still an issue. I went ahead and just squashed all commits.

**Comment by patrick96** [comment]
Done! Thanks a lot for all your efforts :smiley:

**Comment by Filip62** [comment]
Thanks a lot for all YOUR efforts.

Also, if you don't mind:


> If we remove the `backslash_pos == value.size() - 1` check, why would it lead to out of bounds access?
> Wouldn't it be just looking at the string termination character?

**Comment by patrick96** [comment]
C++ strings don't have a termination character. If you do str[str.size()], you will access one element to the right of the last one.
The `std::string` class stores the string size separately and doesn't need a termination character.

**Comment by Filip62** [comment]
I should've known that, I know the size is stored internally, but no one ever told me that the null termination is not there in C++, I am amazed and embarrassed.

**Pull request merged**
